### PR TITLE
Add new test cases for the syncfiles feature

### PR DESCRIPTION
**Pull request by neo954**
This test case is for task xcat2/xcat2-task-management#119

### CODE REVIEW

**Comment by hu-weihua** [comment]
Does all the test scenario in [Test Design of syncfiles](https://github.com/xcat2/xcat-core/wiki/Test-Design-of-syncfiles) have been implemented?

**Comment by hu-weihua** [comment]
@neo954 could you pasted the UT result in this PR?

**Comment by hu-weihua** [comment]
There is no one debug message in whole test case. If this case failed, we will have no idea about what happened. Could you make this case can be debug  in the log? thanks

**Comment by immarvin** [comment]
hi @neo954 , did you cover the case on `EXECUTEALWAYS` in syncfiles?

**Comment by neo954** [comment]
```
# xcattest -t updatenode_syncfiles -f foo.conf
xCAT automated test started at Tue May 22 04:49:45 2018
******************************
loading Configure file
******************************
Varible:
    CN = c910f03c01p10
******************************
Initialize xCAT test environment by definition in configure file
******************************
******************************
To detect current test environment
******************************
Detecting: OS = linux
Detecting: ARCH = ppc64le
Detecting: HCP = hmc
******************************
loading test cases
******************************
To run:
updatenode_syncfiles
******************************
Start to run test cases
******************************
------START::updatenode_syncfiles::Time:Tue May 22 04:49:47 2018------

RUN:/opt/xcat/share/xcat/tools/autotest/testcase/updatenode/syncfiles c910f03c01p10 [Tue May 22 04:49:47 2018]
ElapsedTime:27 sec
RETURN rc = 4
OUTPUT:
+ COMPUTE_NODE=c910f03c01p10
++ lsdef c910f03c01p10 -i provmethod
++ awk -F = '/provmethod=/ { print $2 }'
+ OSIMAGE=rhels7.5-alternate-ppc64le-netboot-cudafull
++ mktemp -d /tmp/syncfiles.XXXXXXXX
+ TMP_DIR=/tmp/syncfiles.1q3BcjLz
+ trap cleanup EXIT
+ declare -i failed=0
+ mkdir -p /tmp/syncfiles.1q3BcjLz/files
+ echo foo
+ echo bar
+ echo baz
+ cat
+ chdef -t osimage rhels7.5-alternate-ppc64le-netboot-cudafull synclists=/tmp/syncfiles.1q3BcjLz/foobar.synclist
1 object definitions have been created or modified.
+ updatenode c910f03c01p10 -F
File synchronization has completed for nodes: "c910f03c01p10"
+ ((  failed += 0  ))
+ xdsh c910f03c01p10 'grep foo /etc/files/foo'
c910f03c01p10: foo
+ ((  failed += 0  ))
+ xdsh c910f03c01p10 'grep bar /etc/files/bar'
c910f03c01p10: bar
+ ((  failed += 0  ))
+ xdsh c910f03c01p10 'grep baz /etc/files/baz'
c910f03c01p10: baz
+ ((  failed += 0  ))
+ compute_node_cleanup
+ xdsh c910f03c01p10 'rm -rf /etc/files'
+ updatenode c910f03c01p10 -F -r /usr/bin/scp
File synchronization has completed for nodes: "c910f03c01p10"
+ ((  failed += 0  ))
+ xdsh c910f03c01p10 'grep foo /etc/files/foo'
c910f03c01p10: foo
+ ((  failed += 0  ))
+ xdsh c910f03c01p10 'grep bar /etc/files/bar'
c910f03c01p10: bar
+ ((  failed += 0  ))
+ xdsh c910f03c01p10 'grep baz /etc/files/baz'
c910f03c01p10: baz
+ ((  failed += 0  ))
+ compute_node_cleanup
+ xdsh c910f03c01p10 'rm -rf /etc/files'
+ xdsh c910f03c01p10 'mkdir -p /etc/files/bar'
+ updatenode c910f03c01p10 -F
File synchronization has completed for nodes: "c910f03c01p10"
+ ((  failed += 0  ))
+ xdsh c910f03c01p10 'grep foo /etc/files/foo'
c910f03c01p10: foo
+ ((  failed += 0  ))
+ xdsh c910f03c01p10 'grep bar /etc/files/bar'
c910f03c01p10: bar
+ ((  failed += 0  ))
+ xdsh c910f03c01p10 'grep baz /etc/files/baz'
c910f03c01p10: baz
+ ((  failed += 0  ))
+ compute_node_cleanup
+ xdsh c910f03c01p10 'rm -rf /etc/files'
+ xdsh c910f03c01p10 'mkdir -p /etc/files/bar'
+ updatenode c910f03c01p10 -F -r /usr/bin/scp
c910f03c01p10: scp: /etc/files/bar: Is a directory
File synchronization failed for nodes: "c910f03c01p10"
+ ((  failed += 1  ))
+ xdsh c910f03c01p10 'grep foo /etc/files/foo'
c910f03c01p10: foo
+ ((  failed += 0  ))
+ xdsh c910f03c01p10 'grep bar /etc/files/bar'
c910f03c01p10: grep: /etc/files/bar: Is a directory
+ ((  failed += 1  ))
+ xdsh c910f03c01p10 'grep baz /etc/files/baz'
c910f03c01p10: baz
+ ((  failed += 0  ))
+ compute_node_cleanup
+ xdsh c910f03c01p10 'rm -rf /etc/files'
+ rm -f /tmp/syncfiles.1q3BcjLz/files/bar
+ updatenode c910f03c01p10 -F
c910f03c01p10: rsync: link_stat "/tmp/syncfiles.1q3BcjLz/files/bar" failed: No such file or directory (2)
c910f03c01p10: rsync error: some files/attrs were not transferred (see previous errors) (code 23) at main.c(1178) [sender=3.1.2]
File synchronization failed for nodes: "c910f03c01p10"
+ xdsh c910f03c01p10 'grep foo /etc/files/foo'
c910f03c01p10: foo
+ ((  failed += 0  ))
+ xdsh c910f03c01p10 'grep baz /etc/files/baz'
c910f03c01p10: baz
+ ((  failed += 0  ))
+ compute_node_cleanup
+ xdsh c910f03c01p10 'rm -rf /etc/files'
+ updatenode c910f03c01p10 -F -r /usr/bin/scp
c910f03c01p10: /tmp/syncfiles.1q3BcjLz/files/bar: No such file or directory
File synchronization failed for nodes: "c910f03c01p10"
+ xdsh c910f03c01p10 'grep foo /etc/files/foo'
c910f03c01p10: foo
+ ((  failed += 0  ))
+ xdsh c910f03c01p10 'grep baz /etc/files/baz'
c910f03c01p10: baz
+ ((  failed += 0  ))
+ compute_node_cleanup
+ xdsh c910f03c01p10 'rm -rf /etc/files'
+ mkdir -p /tmp/syncfiles.1q3BcjLz/files/bar
+ xdsh c910f03c01p10 'mkdir -p /etc/files && echo bar >/etc/files/bar'
+ updatenode c910f03c01p10 -F
File synchronization has completed for nodes: "c910f03c01p10"
+ ((  failed += 0  ))
+ xdsh c910f03c01p10 'grep foo /etc/files/foo'
c910f03c01p10: foo
+ ((  failed += 0  ))
+ xdsh c910f03c01p10 'test -d /etc/files/bar'
+ ((  failed += 0  ))
+ xdsh c910f03c01p10 'grep baz /etc/files/baz'
c910f03c01p10: baz
+ ((  failed += 0  ))
+ compute_node_cleanup
+ xdsh c910f03c01p10 'rm -rf /etc/files'
+ xdsh c910f03c01p10 'mkdir -p /etc/files && echo bar >/etc/files/bar'
+ updatenode c910f03c01p10 -F -r /usr/bin/scp
c910f03c01p10: scp: /etc/files/bar: Not a directory
File synchronization failed for nodes: "c910f03c01p10"
+ ((  failed += 1  ))
+ xdsh c910f03c01p10 'grep foo /etc/files/foo'
c910f03c01p10: foo
+ ((  failed += 0  ))
+ xdsh c910f03c01p10 'test -d /etc/files/bar'
+ ((  failed += 1  ))
+ xdsh c910f03c01p10 'grep baz /etc/files/baz'
c910f03c01p10: baz
+ ((  failed += 0  ))
+ compute_node_cleanup
+ xdsh c910f03c01p10 'rm -rf /etc/files'
+ exit 4
+ cleanup
+ rm -rf /tmp/syncfiles.1q3BcjLz
+ chdef -t osimage rhels7.5-alternate-ppc64le-netboot-cudafull synclists=
1 object definitions have been created or modified.
CHECK:rc == 0	[Failed]

------END::updatenode_syncfiles::Failed::Time:Tue May 22 04:50:14 2018 ::Duration::27 sec------
------Total: 1 , Failed: 1------

xCAT automated test finished at Tue May 22 04:50:14 2018
Please check results in the /opt/xcat/bin/../share/xcat/tools/autotest/result/,
and see /opt/xcat/bin/../share/xcat/tools/autotest/result//failedcases.20180522044945 file for failed cases.
see /opt/xcat/bin/../share/xcat/tools/autotest/result//performance.report.20180522044945 file for time consumption
```

**Comment by hu-weihua** [comment]
@neo954 , to be honest, the output of this test case is hard to understand, even reading source code ahead. I strongly recommend refine the output.

**Comment by hu-weihua** [comment]
Need to split this test case to more cases to test ``rsync`` and ``scp`` separately. Because:
* The target is to make this case to be a common case that can be run on any kind of compute node. But in cumulus switch, it does not support ``rsync``, so this case will fail certainly. 
* Due to ``rsync`` and ``scp`` have different processing against nonexistent file or directory, so the test pass standard of these two command are different too.

File: xCAT-test/autotest/testcase/updatenode/syncfiles [hunk 1]
```diff
@@ -54,11 +54,11 @@ compute_node_cleanup
 xdsh "${COMPUTE_NODE}" 'mkdir -p /etc/files/bar'
 updatenode "${COMPUTE_NODE}" -F ${NODE_RCP}
 # Check return code
-(( failed += $? ))
+[ -z "${NODE_RCP}" ] && (( failed += $? ))
```
Comment (hu-weihua) [inline xCAT-test/autotest/testcase/updatenode/syncfiles]: If the method is ``scp``, that means we just run ``updatenode  <node> -r scp`` then do nothing. If a test case just run a command then do nothing, what is the meaning of a test case?

Comment (neo954) [reply xCAT-test/autotest/testcase/updatenode/syncfiles]: Hello @hu-weihua,

What you believe is not true. And you may misunderstand the bash code. While the exit code of the `updatenode` command is checked, that does not mean the command is doing nothing.

In this test case scenario, a small part of the test may fail (when try to replace a directory with a regular file). That is due to the limitation of `updatenode`. This test case is ensure, even in this scenario, the rest of the result is checked (check the content of other copied files). Thus, as a result, partial failure can be tolerated and is acceptable. And this is the design goal of this test case.

**Comment by immarvin** [comment]
agree to merge

**Pull request merged**